core/src/main/java/nl/nn/adapterframework/pipes/TimeoutGuardPipe.java

```diff
@@ -97,39 +61,57 @@ public PipeRunResult doPipe(Message message, IPipeLineSession session) throws Pi
 			timeout_work = Integer.valueOf(timeout_work_str);
 		}
 
-		DoPipe doPipe = new DoPipe(message, session, Thread.currentThread().getName(), ThreadContext.peek());
-		ExecutorService service = Executors.newSingleThreadExecutor();
-		Future<Message> future = service.submit(doPipe);
-		Object result = null;
+		log.debug(getLogPrefix(session) + "setting timeout of [" + timeout_work + "] s");
+		TimeoutGuard tg = new TimeoutGuard(timeout_work, getName()) {
+			@Override
+			protected void kill() {
+				//The guard automatically kills the current thread, additional threads maybe 'killed' by implementing killPipe.
+				killPipe();
+			}
+		};
+
 		try {
-			log.debug(getLogPrefix(session) + "setting timeout of [" + timeout_work + "] s");
-			result = future.get(timeout_work, TimeUnit.SECONDS);
+			return doPipeWithTimeoutGuarded(message, session);
 		} catch (Exception e) {
-			String msg;
-			if (e instanceof TimeoutException) {
-				String errorMsg = getLogPrefix(session) + "exceeds timeout of [" + timeout_work + "] s, interupting";
-				future.cancel(true);
-				msg = e.getClass().getName() + ": " + errorMsg;
-			} else {
-				msg = e.getClass().getName();
-			}
+			String msg = e.getClass().getName();
 
 			if (isThrowException()) {
 				throw new PipeRunException(this, msg, e);
 			} else {
 				String msgString = msg + ": " + e.getMessage();
 				log.error(msgString, e);
 				String msgCdataString = "<![CDATA[" + msgString + "]]>";
-				result = "<error>" + msgCdataString + "</error>";
+				Message errorMessage = new Message("<error>" + msgCdataString + "</error>");
+				return new PipeRunResult(getForward(), errorMessage);
 			}
 		} finally {
-			service.shutdown();
+			if(tg.cancel()) {
+				//Throw a TimeOutException
+				String msgString = "TimeOutException";
+				Exception e = new TimeOutException("exceeds timeout of [" + timeout_work + "] s, interupting");
+				if (isThrowException()) {
+					throw new PipeRunException(this, msgString, e);
+				} else {
+					//This is used for the old console, where a message is displayed
+					log.error(msgString, e);
+					String msgCdataString = "<![CDATA[" + msgString + ": "+ e.getMessage() + "]]>";
+					Message errorMessage = new Message("<error>" + msgCdataString + "</error>");
+					return new PipeRunResult(getForward(), errorMessage);
+				}
+			}
 		}
-		return new PipeRunResult(getForward(), result);
 	}
 
-	public Message doPipeWithTimeoutGuarded(Message input, IPipeLineSession session) throws PipeRunException {
-		return input;
+	/**
+	 * doPipe wrapped around a TimeoutGuard
+	 */
+	public abstract PipeRunResult doPipeWithTimeoutGuarded(Message input, IPipeLineSession session) throws PipeRunException;
+
+	/**
+	 * optional implementation to kill additional threads if the pipe may have created those.
+	 */
+	protected void killPipe() {
+		//kill other threads
```

Comment:
killPipe() moet iets hebben waardoor hij weet wat hij moet killen: een class of een handle die je meegeeft bij het creeren van de TimeoutGuard